File: real-time/file.ts
import WebSocket from 'ws';

const socket = new WebSocket('wss://api.assemblyai.com/v2/realtime/ws');

socket.on('open', () => {
  const authHeader = {
    Authorization: '{your_api_token}',
  };

  const sampleRate = 16000;
  const wordBoost = ['HackerNews', 'Twitter'];

  socket.send(
    `wss://api.assemblyai.com/v2/realtime/ws?sample_rate=${sampleRate}&word_boost=${JSON.stringify(
      wordBoost
    )}`,
    { headers: authHeader }
  );
});

socket.on('message', (data: string) => {
  const message = JSON.parse(data);
  if (message.message_type === 'SessionBegins') {
    const sessionId = message.session_id;
    const expiresAt = message.expires_at;
    console.log(`Session ID: ${sessionId}`);
    console.log(`Expires at: ${expiresAt}`);
  } else if (message.message_type === 'PartialTranscript') {
    console.log(`Partial transcript received: ${message.text}`);
  } else if (message.message_type === 'FinalTranscript') {
    console.log(`Final transcript received: ${message.text}`);
  }
});

socket.on('error', (error) => {
  const errorObject = JSON.parse(error);
  if (errorObject.error_code === 4002) {
    console.log('Insufficient Funds');
  }
});

socket.on('close', () => {
  console.log('WebSocket closed');
});

function sendAudio(ws: WebSocket, audioData: ArrayBuffer) {
  const payload = {
    audio_data: Buffer.from(audioData).toString('base64'),
  };
  ws.send(JSON.stringify(payload));
}

function terminateSession(socket: WebSocket) {
  const payload = { terminate_session: true };
  socket.send(JSON.stringify(payload));
  socket.close();
}

File: real-time/file.rb
require 'websocket-client-simple'
require 'base64'
require 'json'

auth_header = {'Authorization': '{your_api_token}'}
sample_rate = 16000
word_boost = ["HackerNews", "Twitter"]
url = "wss://api.assemblyai.com/v2/realtime/ws?word_boost=#{word_boost.to_json}&sample_rate=#{sample_rate}"
ws = WebSocket::Client::Simple.connect url, headers: auth_header

ws.on :message do |msg|
  message = JSON.parse(msg.data)
  case message['message_type']
  when 'SessionBegins'
    session_id = message['session_id']
    expires_at = message['expires_at']
    puts "Session ID: #{session_id}, Expires At: #{expires_at}"
  when 'PartialTranscript'
    text = message['text']
    puts "Partial transcript received: #{text}"
  when 'FinalTranscript'
    text = message['text']
    puts "Final transcript received: #{text}"
  end
end

ws.on :error do |error|
  error_object = JSON.parse(error)
  if error_object['error_code'] == 1013
    puts "Temporary server condition forced blocking client's request"
  end
end

ws.on :close do |e|
  puts 'WebSocket closed'
end

def send_audio(socket, audio_data)
  payload = {
    "audio_data" => Base64.strict_encode64(audio_data)
  }
  socket.send(payload.to_json)
end

def terminate_session(socket)
  payload = {"terminate_session" => true}
  message = payload.to_json
  socket.send(message)
  socket.close
end
